### ISSUE TITLE
ALE explain(): IndexError: index 1 is out of bounds for axis 0 with size 1

### ISSUE THREAD

**Issue by chicodelarosa** [SeldonIO/alibi#319]
I am working on a classification problem where the input has several features and the output is either 0 or 1. After training an XGBoost I do the following calls:

`ale = alibi.explainers.ale.ALE(clf.predict_proba, feature_names = X_test.columns.values, target_names = c_names)`
`explanations = ale.explain(X_test.values, min_bin_points = min_n_bins)`

However, when running the last line above to get the explanations, I get the following error:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-24-c0b5a64fd4fa> in <module>
----> 1 explanations = ale.explain(X_test.values, min_bin_points = min_n_bins)

~/.../alibi/explainers/ale.py in explain(self, X, min_bin_points)
     87                 X=X,
     88                 feature=feature,
---> 89                 min_bin_points=min_bin_points
     90             )
     91             deciles = get_quantiles(X[:, feature], num_quantiles=11)

~/.../alibi/explainers/ale.py in ale_num(predictor, X, feature, min_bin_points)
    300     z_high = X.copy()
    301     z_low[:, feature] = q[indices - 1]
--> 302     z_high[:, feature] = q[indices]
    303     p_low = predictor(z_low)
    304     p_high = predictor(z_high)

IndexError: index 1 is out of bounds for axis 0 with size 1
```

What could be causing this error? Could that be that there are missing values (NaN) in the input? Thanks.

**Comment by jklaise** [comment]
@chicodelarosa that is certainly a possibility as `q` is the quantile of the input feature and `indices` index the quantile for which each observation falls into, I'm not sure it can handle NaN values. The easiest way to check would be to remove or impute the NaN values and run the explainer again.

**Comment by chicodelarosa** [comment]
> @chicodelarosa that is certainly a possibility as `q` is the quantile of the input feature and `indices` index the quantile for which each observation falls into, I'm not sure it can handle NaN values. The easiest way to check would be to remove or impute the NaN values and run the explainer again.

I agree, I was going through the internals of the implementation and saw it. Also, I just checked and after filling the missing values, I am still having the error.

**Comment by karthik1290** [comment]
I am also facing this same error. No NaN values. When I tried to use Boston train data it works. I think it can't handle all train data arrays or is it the binning?

**Comment by jklaise** [comment]
That's worrying, sounds like it would indeed be an issue with the binning. Could you run it with a debugger and see what values of `q` and `indices` are? This probably only happens for a particular feature. Alternatively, can you call `q, num_points = adaptive_grid(values, min_bin_points)` with `values` being the sliced feature vector (i.e. `X[:, feature]`) for every `feature` and inspect the `q`? I feel like what is happening is that quantiles with the specified `min_bin_points` can't be found for some feature which results in some degenerate `q`.

**Comment by jklaise** [comment]
@chicodelarosa @karthik1290 Did you manage to look more into this? I think it's to do with your data and the explainer code is probably unable to find appropriate quantiles. This can happen with features which only take on a few values, e.g. if you have categorical features present which are not supported by ALE.

**Comment by franzoni315** [comment]
@jklaise I was doing some tests and it looks like the problem is when the feature is constant for all rows. 
* The vector `q` becomes: array([0.])
* But the `indices` is a vector filled with 1's:  array([1, 1, 1, 1, 1,...])

**Comment by jklaise** [comment]
@franzoni315 thanks for the insight, I agree we should do some graceful handling with degenerate features like that, although one would expect to not have a constant feature in the modelling stage. I also need to check further if an issue like this could arise even if the feature is not constant.

Agreed on the extended functionality for specifying select features, this would also indirectly alleviate this problem.

**Comment by jklaise** [comment]
As an update, there are a few ways of tackling the issue of constant features. One is to just say that there is no grid, then we can define and return an ALE value of zero (one could think of this as a result of a zero-length interval in which both endpoints coincide). But this is not very satisfying and may mask the underlying relationship. E.g. if I have a model f(x,y) = x + y and but y is constant, both ALE(x) and ALE(y) should still be linear functions with slope equal to one. So as an alternative if y is constant, we could define ALE(y) on a small interval (y-eps, y+eps) to show the linear relationship. The interval should be small because we don't want to extrapolate to regions where we have no data. I think this is a more natural solution to the problem, the only question is how to choose this small interval.

**Comment by jklaise** [comment]
Closing as original problem is resolved by #325 . Feature selection functionality was also added in #324 .
